staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+            "The unstaked balance is not yet available due to unstaking delay"
+        );
+        account.unstaked -= amount;
+        self.save_account(&account_id, account);
+        Promise::new(account_id).transfer(amount);
+
+        if self.total_staked_balance > self.last_locked_account_balance {
+            self.restake();
+        }
+    }
+
+    /// Stakes previously deposited money by given account on this account.
+    pub fn stake(&mut self, amount: U128) {
+        self.ping();
+
+        let amount: Balance = amount.into();
+        assert!(amount > 0, "Staking amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+
+        let num_shares = if self.total_staked_balance > 0 {
+            // price = total_staked / total_shares
+            // Price is fixed
+            // (total_staked + amount) / (total_shares + num_shares) = total_staked / total_shares
+            // (total_staked + amount) * total_shares = total_staked * (total_shares + num_shares)
+            // amount * total_shares = total_staked * num_shares
+            // num_shares = amount * total_shares / total_staked
+            // NOTE: Number of shares the account gets is rounded up, but the gas rebate within
+            // contract will compensate for the rounding errors.
+            // Rounding up `a / b` using `(a + b - 1) / b`.
+            ((U256::from(self.total_staked_shares) * U256::from(amount)
+                + U256::from(self.total_staked_balance - 1))
+                / U256::from(self.total_staked_balance))
+            .as_u128()
+        } else {
+            amount
+        };
+
+        assert!(
+            account.unstaked >= amount,
+            "Not enough unstaked balance to stake"
+        );
+        account.unstaked -= amount;
+        account.staked_shares += num_shares;
+        self.save_account(&account_id, account);
+
+        self.total_staked_balance += amount;
+        self.total_staked_shares += num_shares;
+
+        self.restake();
+    }
+
+    /// Request withdrawal for epoch + 2 by sending unstaking transaction for
+    /// `current locked - (given account deposit + rewards)`
+    pub fn unstake(&mut self, amount: U128) {
+        self.ping();
+
+        let amount: Balance = amount.into();
+        assert!(amount > 0, "Unstaking amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+
+        assert!(
+            self.total_staked_balance > 0,
+            "The contract doesn't have staked balance"
+        );
+        // NOTE: The number of shares the account will pay is rounded up, to avoid giving extra
+        // amount.
+        // Rounding up `a / b` using `(a + b - 1) / b`.
+        let num_shares = ((U256::from(self.total_staked_shares) * U256::from(amount)
```

Comment:
put shares computation into separate function as it's the same as line 165 and can have a nice comment there.
also can then test separately (in module level tests can set private fields directly).